sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetVectorUpdater.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet;
+
+import org.apache.parquet.column.Dictionary;
+
+import org.apache.spark.sql.execution.vectorized.WritableColumnVector;
+
+public interface ParquetVectorUpdater {
+  /**
+   * Read a batch of `total` values from `valuesReader` into `values`, starting from `offset`.
+   *
+   * @param total total number of values to read
+   * @param offset starting offset in `values`
+   * @param values destination values vector
+   * @param valuesReader reader to read values from
+   */
+  void updateBatch(
+      int total,
+      int offset,
+      WritableColumnVector values,
+      VectorizedValuesReader valuesReader);
+
+  /**
+   * Read a single value from `valuesReader` into `values`, at `offset`.
+   *
+   * @param offset offset in `values` to put the new value
+   * @param values destination value vector
+   * @param valuesReader reader to read values from
+   */
+  void update(int offset, WritableColumnVector values, VectorizedValuesReader valuesReader);
+
+  /**
+   * Decode a batch of `total` dictionary IDs from `dictionaryIds` into `values` starting at
+   * `offset`, using `dictionary`.
+   *
+   * @param total total number of dictionary IDs to decode
+   * @param offset starting offset in `values`
+   * @param values destination vector
+   * @param dictionaryIds vector storing the dictionary IDs
+   * @param dictionary Parquet dictionary used to decode dictionary ID to its value
+   */
+  default void decodeDictionaryIds(
+      int total,
+      int offset,
+      WritableColumnVector values,
+      WritableColumnVector dictionaryIds,
+      Dictionary dictionary) {
+    for (int i = offset; i < offset + total; i++) {
+      if (!values.isNullAt(i)) {
```

Comment:
why do we need this condition? `values` is the destination, isn't it?

Reply:
The nulls in `values` are populated previously in `VectorizedRleValuesReader.readIntegers`. Here we are only decoding non-null values from the dictionary IDs.